pkg/provider-local/webhook/controlplane/ensurer.go

```diff
@@ -84,7 +84,21 @@ func (e *ensurer) EnsureMachineControllerManagerDeployment(ctx context.Context,
 	}
 
 	container := machinecontrollermanager.GetSidecarContainer(newObj.Namespace, "provider-local", image.String())
-	newObj.Spec.Template.Spec.Containers = append(newObj.Spec.Template.Spec.Containers, container)
+
+	// check whether the container has already been injected into the
+	// mcm deployment, as we do not want to append it to the containers
+	// list in that case.
+	containerAlreadyInjected := false
+	for _, c := range newObj.Spec.Template.Spec.Containers {
+		if c.Image == container.Image {
+			containerAlreadyInjected = true
+			break
+		}
+	}
+
+	if !containerAlreadyInjected {
+		newObj.Spec.Template.Spec.Containers = append(newObj.Spec.Template.Spec.Containers, container)
+	}
```

Comment:
You could use https://github.com/gardener/gardener/blob/5abed1bfedd6f2188e5dcec9af161d1b5c286236/extensions/pkg/webhook/utils.go#L213 instead